### PR TITLE
Use safe-io

### PR DESCRIPTION
**Pull request by A6GibKm**
See https://doc.rust-lang.org/src/std/os/fd/owned.rs.html#94

### CODE REVIEW

**Comment by A6GibKm** [comment]
Multiple things need figuring out, for example if pipewire_streams should use OwnedFd instead of BorrowedFd.

**Comment by A6GibKm** [comment]
Also are those unsafe calls Ok? I am sure they will insta crash if the Fd is -1

**Comment by A6GibKm** [comment]
Also requires Rust 1.63

**Comment by A6GibKm** [comment]
Also this one does `fd.close()` on Drop

File: src/desktop/camera.rs [hunk 1]
```diff
@@ -209,13 +212,7 @@ fn pipewire_streams_inner<F: Fn(Stream) + Clone + 'static, G: FnOnce() + Clone +
 /// *Note* The socket referenced by `fd` must not be used while this function is
 /// running.
 #[cfg(feature = "pipewire")]
-pub async fn pipewire_streams(fd: RawFd) -> Result<Vec<Stream>, pw::Error> {
```
Comment (bilelmoussaoui) [inline src/desktop/camera.rs:212]: this means we can drop the dependency on libc right?

Comment (A6GibKm) [reply src/desktop/camera.rs:212]: Yes, dropped.

**Comment by A6GibKm** [comment]
Ported the rest of the modules, the only one that requires extra consideration is flatpak/mod.rs since we store the fds of the builder as RawFd, we could store them as OwnedFd.

**Comment by A6GibKm** [comment]
This MR is missing an update to the demo Cargo.toml.

File: src/flatpak/mod.rs [hunk 2]
```diff
@@ -125,28 +120,28 @@ pub struct SpawnOptions {
     /// A list of file descriptor for files inside the sandbox that will be
     /// exposed to the new sandbox, for reading and writing.
     #[zvariant(rename = "sandbox-expose-fd")]
-    sandbox_expose_fd: Option<Vec<Fd>>,
+    sandbox_expose_fd: Vec<Fd<'f>>,
```
Comment (bilelmoussaoui) [inline src/flatpak/mod.rs:123]: there is a reason why all those are optional, zbus's SerializeDict doesn't provide a skip_serializing_if functionnality and the portals would explode.

Comment (A6GibKm) [reply src/flatpak/mod.rs:123]: Ah ok, I thought None was the same as an empty vec

Comment (A6GibKm) [reply src/flatpak/mod.rs]: Using `Option<Vec<Fd<'f>>>` now.

**Comment by A6GibKm** [comment]
There is also PidFd in rust nigthly, I have a small [commit](https://github.com/A6GibKm/ashpd/commit/caebbe206752251236075dcd28904d8e2160094f) for posterity.

**Comment by bilelmoussaoui** [comment]
This PR contains way too many changes, some of them are still interesting but would need a newer one. So closing this one

**Pull request closed without merging**